src/modules/swap/services/swapFlow/index.ts

```diff
@@ -18,18 +22,80 @@ export async function swapFlow(
   priceImpactParams: PriceImpact,
   confirmPriceImpactWithoutFee: (priceImpact: Percent) => Promise<boolean>
 ) {
+  // TODO
+  const isTokenSupportsEIP2612 = true
+
   logTradeFlow('SWAP FLOW', 'STEP 1: confirm price impact')
   if (priceImpactParams?.priceImpact && !(await confirmPriceImpactWithoutFee(priceImpactParams.priceImpact))) {
     return
   }
 
+  let newAppData = null
+
+  if (isTokenSupportsEIP2612) {
+    const sellTokenContract = input.sellTokenContract
+    const inputToken = input.context.trade.inputAmount.currency
+    const nonce = await sellTokenContract?.nonces(input.orderParams.account)
+
+    const permit = {
+      owner: input.orderParams.account,
+      spender: GP_VAULT_RELAYER[input.orderParams.chainId],
+      value: BigInt(input.context.trade.inputAmount.toExact()).toString(16),
+      nonce: nonce ? nonce.toNumber() : 0,
+      deadline: MaxUint256.toString(),
+    }
+    console.log('GGGGGG', input.orderParams.signer.provider)
+    const permitSignature = splitSignature(
+      await (input.orderParams.signer.provider as any)._signTypedData(
+        {
+          name: inputToken.name,
+          version: 1, // TODO
+          chainId: inputToken.chainId,
+          verifyingContract: (inputToken as Token).address,
+        },
```

Comment:
Unfortunately, there is no standard for getting structured domain data 😞. For example, the COW token is hardcoded as `1`, for `USDC`, there is a `version()` function that you can call.

In fact, you are not required to use all the domain fields, a valid EIP-2612 implementation could just use:

```ts
        {
          name: inputToken.name,
        }
```

Reply:
https://github.com/1inch/permit-signed-approvals-utils
This util covers the majority of tokens

Reply:
https://github.com/1inch/permit-signed-approvals-utils/blob/master/src/eip-2612-permit.const.ts
Here you can find different structures